### ISSUE TITLE
If a reject happens after `.entity()` is called, argument is undefined

### ISSUE THREAD

**Issue by phated** [cujojs/rest#119]
I've been working with rest on a project where I don't care about anything but the entity, so I use the `.entity()` method available on a client's promise.  If the request errors (due to the `errorCode` interceptor in my case), the argument to `.catch()` is undefined.  I've tracked this to https://github.com/cujojs/rest/blob/master/util/responsePromise.js#L22 which attempts to pluck the `entity` property from the response, but the error exists as an `error` property.  Should these helper methods not wire up an error handler to the promise?

**Comment by scothis** [comment]
The errorBack for the helper methods should receive the equivalent property on the rejected response. There's a test specifically for the [case of a rejected entity](https://github.com/cujojs/rest/blob/master/test/util/responsePromise-test.js#L59-L68).

If you skip `.entity()` and instead `.catch()` directly on the response, what is the entity value on the resolved value?

**Comment by phated** [comment]
There isn't an `entity` property on the rejected response object.  The response object is:
```js
{ request:
   { path: 'http://localhost:8000/table',
     originator:
      { [Function: interceptedClient]
        skip: [Function: skip],
        wrap: [Function: wrap],
        chain: [Function: chain] },
     headers:
      { 'Content-Type': 'application/json',
        Accept: 'application/json, application/json;q=0.8, text/plain;q=0.5, */*;q=0.2' },
     method: 'GET',
     canceled: false,
     cancel: [Function: cancel] },
  error:
   { [Error: connect ECONNREFUSED]
     code: 'ECONNREFUSED',
     errno: 'ECONNREFUSED',
     syscall: 'connect' } }
```

Maybe this case is specific to OS level errors such as `ECONNREFUSED`?

**Comment by scothis** [comment]
My instinct is that this works as expected. You asked for the entity and there is no entity.

Since rest.js was created, the best practice around promise rejections has settled on the rejected value being an Error instance. rest.js struggles with what is an exceptional vs normal code path.

Making any change is a major break. I'm open to rethinking how errors should work for 2.0.

**Comment by pdf** [comment]
> Making any change is a major break

Is it though?  Expected behaviour is for the promise to be rejected with a useful result.  Having to work through nested promises makes for unnecessary boilerplate and makes integration with other libraries problematic.

**Comment by bodell** [comment]
I also ran into this error. [https://github.com/cujojs/rest/blob/master/util/responsePromise.js](https://github.com/cujojs/rest/blob/master/util/responsePromise.js)#L20.  What I discovered was that it wasn't that the entity didn't exist, but rather the entity being bound to had a slight error in it causing loading to fail.  However this was extremely unclear given that the error was "undefined".  I was able to implement an easy fix by modifying the line to:
return Promise.reject(value && value[name] || value);  This way, if the name property doesn't exist, you still get an error that has some meaning.  Let me know if you want me to submit a PR for this.
